Examples/WindowsFormsApplication/Nfield.cs

```diff
@@ -0,0 +1,138 @@
+
+            NfieldConnection = NfieldConnectionFactory.Create(ApplicationConfiguration.Current.NfieldApiUri);
+            NfieldConnection.RegisterTokenProvider(domainName, ProvideTokenAsync);
+        }
+
+        private IPublicClientApplication CreateClientApp()
+        {
+            var client = PublicClientApplicationBuilder.Create(ApplicationConfiguration.Current.ClientId)
+                .WithRedirectUri("http://localhost")
+                .WithAuthority(AzureCloudInstance.AzurePublic, ApplicationConfiguration.Current.Tenant)
+                .Build();
+
+            client.UserTokenCache.SetBeforeAccess(BeforeAccessNotification);
+            client.UserTokenCache.SetAfterAccess(AfterAccessNotification);
+
+            return client;
+        }
+
+        private async Task<string> ProvideTokenAsync()
+        {
+            var result = await AuthenticateAsync((status) => { });
+            return result.AccessToken;
+        }
+
+        private static string CreateScope(string scope) => $"{ApplicationConfiguration.Current.NfieldApiApplicationId}/{scope}";
+
+        internal async Task LogoutAsync()
+        {
+            var accounts = (await ClientApp.GetAccountsAsync()).ToList();
+            while (accounts.Any())
+            {
+                await ClientApp.RemoveAsync(accounts.First());
+                accounts = (await ClientApp.GetAccountsAsync()).ToList();
+            }
+        }
+
+        private byte[] RestoreTokenCacheFromRegistry()
+        {
+            using (RegistryKey key = Registry.CurrentUser.CreateSubKey(@"SOFTWARE\NIPO Samples"))
+            {
+                var value = key.GetValue("TokenCache");
+                if (value is byte[] state)
+                    return state;
+            }
+
+            return null;
+        }
+
+        private void SaveTokenCacheInRegistry(byte[] state)
+        {
+            using (RegistryKey key = Registry.CurrentUser.CreateSubKey(@"SOFTWARE\NIPO Samples"))
+            {
+                key.SetValue("TokenCache", state, RegistryValueKind.Binary);
+            }
+        }
+
+        public void BeforeAccessNotification(TokenCacheNotificationArgs args)
+        {
+            args.TokenCache.DeserializeMsalV3(RestoreTokenCacheFromRegistry());
+        }
+
+        public void AfterAccessNotification(TokenCacheNotificationArgs args)
+        {
+            if (args.HasStateChanged)
+            {
+                SaveTokenCacheInRegistry(args.TokenCache.SerializeMsalV3());
+            }
+        }
+
+        public async Task<AuthenticationResult> TryAuthenticateSilentAsync()
+        {
+            var accounts = await ClientApp.GetAccountsAsync();
+            var account = accounts.FirstOrDefault();
+
+            try
+            {
+                return await ClientApp.AcquireTokenSilent(Scopes, account).ExecuteAsync();
+            }
+            catch (MsalUiRequiredException)
+            {
+                return null;
```

Comment:
Not sure how far we want to take this sample but maybe we should add some logging as a best practice :-D

Reply:
I'll add a comment, I don't intend to make this a best practise app all round, just the required pieces to get AAD authentication and authorization up and running.